### PR TITLE
entproto: implement BatchCreate method

### PR DESCRIPTION
**Pull request by JeremyV2014**
Implementation of BatchCreate method for ent/ent#2446

### CODE REVIEW

File: entproto/service.go [hunk 1]
```diff
@@ -30,6 +30,9 @@ const (
 	// MaxPageSize is the maximum page size that can be returned by a List call. Requesting page sizes larger than
 	// this value will return, at most, MaxPageSize entries.
 	MaxPageSize = 1000
+	// MaxBatchCreateSize is the maximum number of entries that can be created by a single BatchCreate call. Requests
+	// exceeding this batch size will return an error.
+	MaxBatchCreateSize = 1000
```
Comment (JeremyV2014) [inline entproto/service.go:35]: Wasn't sure if we wanted granular control (setting for each batch operation) here or one batch size limit for all operations.

File: entproto/cmd/protoc-gen-entgrpc/template/method_batch_create.tmpl [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    {{- $reqVar := camel .G.EntType.Name -}}
+    requests := req.GetRequests()
+    if len(requests) > entproto.MaxBatchCreateSize {
+        return nil, {{ statusErrf "InvalidArgument" "batch size cannot be greater than %d" "entproto.MaxBatchCreateSize" }}
+    }
+    bulk := make([]*ent.{{ .G.EntType.Name }}Create, len(requests))
+    for i, req := range requests {
+        bulk[i] = svc.client.{{ .G.EntType.Name }}.Create()
+        m := bulk[i]
+        {{ $reqVar }} := req.Get{{ .G.EntType.Name }}()
```
Comment (JeremyV2014) [inline entproto/cmd/protoc-gen-entgrpc/template/method_batch_create.tmpl:17]: This section is copied directly from the mutate template. They are different enough that it didn't make sense to put the batch create method into method_mutate.tmpl. However, we could refactor to move this section into its own template that can be called by method_batch_create.tmpl and method_mutate.tmpl.

Comment (rotemtam) [reply entproto/cmd/protoc-gen-entgrpc/template/method_batch_create.tmpl:17]: Yes, feels wrong to duplicate the code both in the templates and then again in the generated code. 

I'm in favor of extracting this part into a private helper function that will be generated if either Create or BatchCreate are wanted by the user and used by both.

File: entproto/internal/todo/ent/proto/entpb/entpb.proto [hunk 3]
```diff
@@ -67,6 +67,14 @@ message ListAttachmentResponse {
   string next_page_token = 2;
 }
 
+message BatchCreateAttachmentsRequest {
+  repeated CreateAttachmentRequest requests = 1;
```
Comment (rotemtam) [inline entproto/internal/todo/ent/proto/entpb/entpb.proto:71]: 😍

File: entproto/service.go (hunk 1 above)
Comment (rotemtam) [reply entproto/service.go:35]: This is a good start, we can add this in a small PR later.

File: entproto/service.go [hunk 4]
```diff
@@ -40,8 +43,10 @@ const (
 	MethodDelete
 	// MethodList generates a List gRPC service method for the entproto.Service.
 	MethodList
+	// MethodBatchCreate generates a Batch Create gRPC service method for the entproto.Service.
+	MethodBatchCreate
 	// MethodAll generates all service methods for the entproto.Service. This is the same behavior as not including entproto.Methods.
-	MethodAll = MethodCreate | MethodGet | MethodUpdate | MethodDelete | MethodList
+	MethodAll = MethodCreate | MethodGet | MethodUpdate | MethodDelete | MethodList | MethodBatchCreate
```
Comment (rotemtam) [inline entproto/service.go:49]: 😍

File: entproto/service.go [hunk 5]
```diff
@@ -121,12 +127,20 @@ func (a *Adapter) genMethodProtos(genType *gen.Type, m Method) (methodResources,
...
 	protoEnumFieldType := descriptorpb.FieldDescriptorProto_TYPE_ENUM
+	repeatedFieldLabel := descriptorpb.FieldDescriptorProto_LABEL_REPEATED
 	singleMessageField := &descriptorpb.FieldDescriptorProto{
 		Name:     strptr(snake(genType.Name)),
 		Number:   int32ptr(1),
 		Type:     &protoMessageFieldType,
 		TypeName: &genType.Name,
 	}
+	repeatedMessageField := &descriptorpb.FieldDescriptorProto{
+		Name:     strptr(snake(genType.Name) + "s"),
```
Comment (rotemtam) [inline entproto/service.go:138]: use inflect.Pluralize as well?

**Review by rotemtam — CHANGES_REQUESTED** [review]
Great start! Thanks so much for working on this. 

Please add a test case where a special pluralization case is invoked to make sure the templates work nicely with the generated code.

File: entproto/internal/todo/ent/schema/pony.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+package schema
+
+import (
+	"entgo.io/contrib/entproto"
+	"entgo.io/ent"
+	"entgo.io/ent/schema"
+	"entgo.io/ent/schema/field"
+)
+
+type Pony struct {
```
Comment (rotemtam) [inline entproto/internal/todo/ent/schema/pony.go:24]:
```suggestion
// Pony is the schema of the Pony entity. It is used to verify the pluralization of the entity name
// in the generated protos and service code, expecting Ponies and not Ponys.
type Pony struct {
```

File: entproto/internal/todo/ent/proto/entpb/entpb_user_service.go [hunk 7]
```diff
@@ -137,6 +137,19 @@ func toProtoUser(e *ent.User) (*User, error) {
 	return v, nil
 }
 
+// toProtosUser transforms a list of ent type to a list of pb type
+func toProtosUser(e []*ent.User) ([]*User, error) {
```
Comment (rotemtam) [inline entproto/internal/todo/ent/proto/entpb/entpb_user_service.go:141]:
```suggestion
func toProtoUserList(e []*ent.User) ([]*User, error) {
```

File: entproto/internal/todo/ent/proto/entpb/entpb_user_service.go [hunk 8]
```diff
@@ -449,20 +462,126 @@ func (svc *UserService) List(ctx context.Context, req *ListUserRequest) (*ListUs
...
+
+// BatchCreate implements UserServiceServer.BatchCreate
+func (svc *UserService) BatchCreate(ctx context.Context, req *BatchCreateUsersRequest) (*BatchCreateUsersResponse, error) {
+	requests := req.GetRequests()
+	if len(requests) > entproto.MaxBatchCreateSize {
+		return nil, status.Errorf(codes.InvalidArgument, "batch size cannot be greater than %d", entproto.MaxBatchCreateSize)
+	}
+	bulk := make([]*ent.UserCreate, len(requests))
+	for i, req := range requests {
+		user := req.GetUser()
```
Comment (rotemtam) [inline entproto/internal/todo/ent/proto/entpb/entpb_user_service.go:487]: can this generated code be extracted to a generated function that is used by both Create and BatchCreate?

**Comment by JeremyV2014** [comment]
@rotemtam , I think I'm all done!

File: entproto/internal/todo/ent/proto/entpb/entpb_attachment_service.go [hunk 9]
```diff
@@ -222,20 +231,64 @@ func (svc *AttachmentService) List(ctx context.Context, req *ListAttachmentReque
...
+		return nil, status.Errorf(codes.AlreadyExists, "already exists: %s", err)
+	case ent.IsConstraintError(err):
+		return nil, status.Errorf(codes.InvalidArgument, "invalid argument: %s", err)
+	default:
+		return nil, status.Errorf(codes.Internal, "internal error: %s", err)
+	}
+
+}
+
+func (svc *AttachmentService) configureCreateBuilder(attachment *Attachment) (*ent.AttachmentCreate, error) {
```
Comment (rotemtam) [inline entproto/internal/todo/ent/proto/entpb/entpb_attachment_service.go:283]:
```suggestion
func (svc *AttachmentService) createBuilder(attachment *Attachment) (*ent.AttachmentCreate, error) {
```

**Review by rotemtam — APPROVED** [review]
@JeremyV2014 ,

Great job. One final request and we're good - if we can rename `configureCreateBuilder` to `createBuilder` or simply `builder`, the configure doesn't add any value IMO.

**Comment by rotemtam** [comment]
Amazing work @JeremyV2014 !

**Comment by a8m** [comment]
Thanks, @JeremyV2014 and @rotemtam for adding this.

Please, don't forget to update the docs 🙏

File: entproto/cmd/protoc-gen-entgrpc/main.go [hunk 10]
```diff
@@ -107,6 +108,7 @@ func (g *serviceGenerator) generate() error {
 			"entIdent":     g.entIdent,
 			"newConverter": g.newConverter,
 			"unquote":      strconv.Unquote,
+			"pluralize":    inflect.Pluralize,
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/main.go:111]: This function already exists in the templates functions with support for custom rules. 
https://github.com/ent/ent/blob/master/entc/gen/func.go#L40

cc @rotemtam

**Pull request merged**